[c2goto] avoid clib round-trip via tmp-file & fix `--binary` input handling

fbrausse opened the pull request:
This PR adds an API to read GOTO objects directly from memory and moves loading the internal clib to that new function. This avoids dumping the array to a tmp-file and directly reading it back during parsing; it is implemented using `boost::iostreams` classes. Additionally, the input file handling under `--binary` is fixed and now follows that of ordinary `.c` inputs.

fbrausse committed:
[c2goto] simplify access to internal clibs

fbrausse committed:
[c2goto] flail.py-generated external symbols are const

fbrausse committed:
[goto-programs] make read_goto_binary() fail when read_bin_goto_object() does

fbrausse committed:
[goto-programs] add read_goto_binary() overload reading from a file given by a path

fbrausse committed:
[c2goto] simplify using file_operations API

fbrausse committed:
[c2goto] remove unused #undef

fbrausse committed:
[langapi] language_uit::parse(): avoid a copy

fbrausse committed:
[esbmc] move all users of read_goto_binary() to new version

fbrausse committed:
[goto-programs] remove unused version of read_goto_binary()

fbrausse committed:
[esbmc] fix for handling input files under --binary

fbrausse committed:
[goto-programs c2goto] add_cprover_library(): avoid round-trip via tmp-file

fbrausse committed:
[c2goto] remove macro NO_CPROVER_LIBRARY

fbrausse committed:
[c2goto] remove unused helper function is_in_list()

fbrausse committed:
[c2goto] remove unused variable

fbrausse committed:
add boost-iostreams dependency

lucasccordeiro reviewed (changes requested):
Overall, this PR looks okay to me.

Shall we consider some regression tests to check the `--binary` option?

fbrausse commented:
> Overall, this PR looks okay to me.
> 
> Shall we consider some regression tests to check the `--binary` option?

Thanks! #557 is about getting `--binary` to work again, this is not fixed by this PR. What the patch does is treating existence of the `--binary` flag as a switch that interprets all file parameters as GOTO inputs, while before the code wrongly (according to the `--help` message and option definition) assumed the flag had a file as parameter. For now, a regression test could only check that `esbmc --binary input.goto` does *not* output ```Failed to open `'```.

Edit: What happens now is
```
ESBMC version 6.8.0 64-bit x86_64 linux
Reading GOTO program from file
GOTO program creation time: 0.022s
GOTO program processing time: 0.001s
Starting Bounded Model Checking
main symbol not found; please set an entry point
Aborted
```

lucasccordeiro commented:
> For now, a regression test could only check that `esbmc --binary input.goto` does *not* output ```Failed to open `'```.

It should be okay for such a test case.

lucasccordeiro commented:
BTW, once you push the test case, could you also rebase this branch?

fbrausse committed:
[regression] add test that --binary --goto-functions-only works

The input.goto binary for this test has been been created with
$ c2goto input.c --output input.goto

fbrausse commented:
OK, done.

rafaelsamenezes commented on regression/esbmc/github_576_binary/test.desc:
```diff
@@ -0,0 +1,4 @@
+CORE
+input.goto
```
Would you mind adding the file that you used to generate this `input.goto`? This way we can try to create more complex tests and also help refactor if we change the enconding.

fbrausse replied:
Done.

rafaelsamenezes reviewed (changes requested):
Minor request.

lucasccordeiro reviewed (approved):
LGTM.

fbrausse commented:
macOS segfaulted and Windows had some failure I cannot identify:
```
 - gnuwin32-coreutils.portable (exited 404) - Error while running 'C:\ProgramData\chocolatey\lib\gnuwin32-coreutils.portable\tools\chocolateyinstall.ps1'.
```
I'll restart the CI.

mikhailramalho commented on src/c2goto/cprover_library.cpp:
```diff
@@ -149,79 +127,36 @@ void add_cprover_library(contextt &context, const messaget &message_handler)
+    message_handler.warning(
+      "Warning: this version of ESBMC does not have a C library "
+      "for 16 bit machines");
+    return;
```
I wonder if we should keep these checks for 16-bit, do we have any user on it on your side @lucasccordeiro? 

I mean, we don't even have models for 16-bit arch.

lucasccordeiro replied:
> I wonder if we should keep these checks for 16-bit, do we have any user on it on your side @lucasccordeiro?

I don't think so.

@rafaelsamenezes: do you have any thoughts here?

rafaelsamenezes replied:
We can open a new issue about removing them.

mikhailramalho reviewed (approved):
LGTM. Do we have any increased memory usage with this patch? 

A great follow-up patch would be improving the code that imports goto functions: I reckon it's not very smart and sometimes we import more than needed.

fbrausse commented:
Thanks for your reviews!
> Do we have any increased memory usage with this patch?

With this patch it was my intention to not increase memory consumption, but I may have failed: measuring the max RSS (using `time(1)`) over 295 regression tests, the average difference between after and before the patch is +282 KiB with a standard deviation of 568 KiB over the tests; with 94 tests using less memory and 200 more. The range for max RSS is 52MiB..180MiB and the clibs each have about 750KiB. I would have expected it to be about the same since `boost::iostreams::array_source` is not making copies (unlike the non-deprecated std-library istreams). At least disk space consumption is less, as is `sys`-time.

Is there an easier way to test memory consumption to some degree of coverage?

mikhailramalho commented:
I think 282 +/- 568 KiB is negligible considering that we usually target a
15GB memory usage.

Em seg., 3 de jan. de 2022 às 15:19, Franz Brauße ***@***.***>
escreveu:

> Thanks for your reviews!
>
> Do we have any increased memory usage with this patch?
>
> With this patch it was my intention to not increase memory consumption,
> but I may have failed: measuring the max RSS (using time(1)) over 295
> regression tests, the average difference between after and before the patch
> is +282 KiB with a standard deviation of 568 KiB over the tests; with 94
> tests using less memory and 200 more. The range for max RSS is
> 52MiB..180MiB and the clibs each have about 750KiB. I would have expected
> it to be about the same since boost::iostreams::array_source is not
> making copies (unlike the non-deprecated std-library istreams). At least
> disk space consumption is less, as is sys-time.
>
> Is there an easier way to test memory consumption to some degree of
> coverage?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/576#issuecomment-1004269736>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH744EDJLJ5DDGPDGUTUUHSBJANCNFSM5LDTL6HA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>


-- 

Mikhail Ramalho.

fbrausse commented:
> A great follow-up patch would be improving the code that imports goto functions: I reckon it's not very smart and sometimes we import more than needed.

As you've probably seen in #557, the `c2goto` program at the moment does not export the goto-functions, so this code path is not used atm. But I'd assume we need those to solve the issue and be happy to take a look at them.

mikhailramalho commented:
Em seg., 3 de jan. de 2022 às 16:16, Franz Brauße ***@***.***>
escreveu:

> A great follow-up patch would be improving the code that imports goto
> functions: I reckon it's not very smart and sometimes we import more than
> needed.
>
> As you've probably seen in #557
> <https://github.com/esbmc/esbmc/issues/557>, the c2goto program at the
> moment does not export the goto-functions, so this code path is not used
> atm. But I'd assume we need those to solve the issue and be happy to take a
> look at them.
>

I was actually talking about the symbol_deps stuff, I see some unnecessary
functions being pulled from time to time when printing the goto functions
(even for code without function pointers that we can't statically determine
where it's pointing to).

> —
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/576#issuecomment-1004300949>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJHZKAPRD5XYUIWGWAUDUUHYZPANCNFSM5LDTL6HA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>


-- 

Mikhail Ramalho.

fbrausse commented:
> I was actually talking about the symbol_deps stuff, I see some unnecessary functions being pulled from time to time when printing the goto functions (even for code without function pointers that we can't statically determine where it's pointing to).

Ack, maybe it's worth opening an issue for that.
